### PR TITLE
ScriptedItems

### PR DESCRIPTION
**Pull request by memmaker**
Items with a dedicated "use" script hook and fully configurable actions.
 - added new BattleType 12 BT_SCRIPTED
 - added script hook scriptedItemUse, called after an user clicks on the action menu item
 - added script hook psiAttack, called after the psi attack success calculation, allows override
 - scriptedItems support up to 5 scriptable actions
 - allowed scripted items to act as firearms
 - allowed scripted items to have actions with zero TU cost, if mana is used instead
 - extended action menu display, zero TU actions will display their mana cost
 - allowed scripted items to act as psi amp
 - allowed scripted items to target friendly units

### CODE REVIEW

File: src/Battlescape/BattlescapeGame.cpp [hunk 1]
```diff
@@ -95,7 +95,7 @@ void BattleActionCost::clearTU()
  */
 bool BattleActionCost::haveTU(std::string *message)
 {
-	if (Time <= 0)
+	if (Time <= 0 && Mana <= 0)
```
Comment (MeridianOXC) [inline src/Battlescape/BattlescapeGame.cpp]: All imaginable actions must cost at least 1 TU. Xcom cannot move or act faster than light. Please remove this sub-feature.

File: src/Mod/ModScript.h [hunk 2]
```diff
@@ -313,6 +321,7 @@ class ModScript
 		ReactionUnitAction,
 		ReactionUnitReaction,
 
+		PsiAttack,
```
Comment (MeridianOXC) [inline src/Mod/ModScript.h]: is this only for the new battle type? or also for normal psi amps? if also for normal psi amps, then please split it into a separate commit... don't put multiple new features into the same commit please

File: src/Battlescape/ActionMenuState.cpp [hunk 3]
```diff
@@ -180,6 +180,44 @@ ActionMenuState::ActionMenuState(BattleAction *action, int x, int y) : _action(a
 	{
 		addItem(BA_USE, "STR_USE_MIND_PROBE", &id, Options::keyBattleActionItem1);
 	}
+	else if (weapon->getBattleType() == BT_SCRIPTED)
+	{
+		if (weapon->getCostUse().Time > 0 || weapon->getCostUse().Mana > 0)
+		{
+			addItem(BA_USE, "STR_USE_SCRIPTED", &id, Options::keyBattleActionItem1);
+		}
+		else if (weapon->getCostSnap().Time > 0 || weapon->getCostSnap().Mana > 0)
```
Comment (MeridianOXC) [inline src/Battlescape/ActionMenuState.cpp]: why else if? I'm pretty sure "fully configurable actions" for modders means they can make a weapon with any action combination they like. With these else ifs, it looks like they are stuck with normal already existing firearms (aimed, snap, auto) or normal already existing psi amps (use, panic, MC)

**Comment by MeridianOXC** [comment]
is there a test mod to demonstrate/test what's new?

File: src/Battlescape/BattlescapeGame.cpp (hunk 1 above)
Comment (memmaker) [reply src/Battlescape/BattlescapeGame.cpp]: Does this also apply to action which are not actually carried out?
There are several possible features I could think of which would justify a 0 TU action game wise. When someone is not acting in that moment, for example.

File: src/Battlescape/ActionMenuState.cpp (hunk 3 above)
Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: I was under the impression, that certain limitations did apply about the slots for the actions. Maybe I misunderstood that part and this could be made even more flexible. Currently it allows for example a PSI Action, a firearm action and a general use action. They are not stuck, because the script can catch each of the actions and re-define them on the fly. That was the idea at least.

**Comment by memmaker** [comment]
[ReportMod.zip](https://github.com/MeridianOXC/OpenXcom/files/4142022/ReportMod.zip)
This includes a ruleset that shows some of the possibilities

File: src/Mod/ModScript.h (hunk 2 above)
Comment (memmaker) [reply src/Mod/ModScript.h]: Will do

File: src/Battlescape/BattlescapeGame.cpp (hunk 1 above)
Comment (MeridianOXC) [reply src/Battlescape/BattlescapeGame.cpp]: 1/ Example please... what soldier skill?
2/ Battleactions are performed on items, not on soldiers... if we want some sort of soldier toggles, they should be handled by a completely new GUI, clicking on the soldier instead of a weapon

File: src/Battlescape/ActionMenuState.cpp (hunk 3 above)
Comment (MeridianOXC) [reply src/Battlescape/ActionMenuState.cpp]: I'm just casually reading the code, didn't investigate much. But on the first look, it looks like you cannot have for example a Use action and a Snapshot action at the same time. True or false?

File: src/Battlescape/BattlescapeGame.cpp (hunk 1 above)
Comment (memmaker) [reply src/Battlescape/BattlescapeGame.cpp]: 1) One example would be "Enter Concealment" allowing the soldier to return to the status before being spotted. Another one would be "Run & Gun" which explicitly refreshed the TU for a soldier so he can do another turn, needing one TU to activate that one is just inconvenient. "Reaper" Skill activation:
A devastating chain melee attack, each melee kill in Reaper mode grants an extra action, but further melee attacks deal reduced damage.

That are just three of the skills I have implemented or plan to implement for the ranger class. Each of these could be justifiably activated with 0 TU.

2) That's why I was eager to have special weapons on soldier types. Because the special weapon icon effectively does what I want.

File: src/Battlescape/ActionMenuState.cpp (hunk 3 above)
Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: False. That is exactly one of the uses cases. The actions as defined here in the menuState can be overridden by the use script. So you could define 5 actions and the script can the choose what these 5 actions are. You could have them all be snapshots or all be use actions, by return the corresponding values. You could also just keep the actions as defined and let them trough. Or you could redefine at will. Leaving the use action be and maybe changing the aimed shot to a snap shot.

Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: I remember now, why I did these if/elses: Options::keyBattleActionItem1. I first thought these specified the position in the menu and so I could not have two actions with the same key. I only later realized, that these are actually just keyboard hotkey bindings. It's still worth thinking about having them bound in a sensible way. I also tried limited the item to 5 actions because of this. But I agree, we could easily have this even more flexible and remove some of the limitations I introduced under a false premise.

File: src/Battlescape/TileEngine.cpp [hunk 4]
```diff
@@ -4106,6 +4115,27 @@ bool TileEngine::medikitUse(BattleAction *action, BattleUnit *target, BattleMedi
...
+	BattleItem *item = action->weapon;
+	BattleActionType actionType = action->type;
+	bool canPerformAction = true;
+	ModScript::ScriptedItemUse::Output args { actionType, canPerformAction };
+	
+	ModScript::ScriptedItemUse::Worker work { action->actor, action->weapon, _save };
+	
+	work.execute(item->getRules()->getScript<ModScript::ScriptedItemUse>(), args);
+	
+	int newActionType = args.getFirst();
```
Comment (memmaker) [inline src/Battlescape/TileEngine.cpp]: This is where the "scriptable" part really comes into play. Even if the action the player selected was BA_USE, the new action returned by the script can be something entirely different.

That leaves the action definitions on the item with a default behavior (the BA that was defined) and a layout option. Firearm actions display an accuracy while the use action does not, for example.

File: src/Battlescape/BattlescapeGame.cpp (hunk 1 above)
Comment (MeridianOXC) [reply src/Battlescape/BattlescapeGame.cpp]: OK, then this is definitely a misuse. Special weapon is also just a weapon/item. Soldier skill activation should not be done via item actions. It needs to be a separate context menu (for example by clicking on the soldier rank/portrait or on some extra icon) that indicates it is a soldier skill.

Comment (memmaker) [reply src/Battlescape/BattlescapeGame.cpp]: What about an icon at the same location of the special weapon icon, that could pop up such a menu? (I have a matching icon for my mod in use.)
If that would be possible, I would not use a weapon/item of course.

Comment (memmaker) [reply src/Battlescape/BattlescapeGame.cpp]: I would propose two different features then:
1. Scripted Items which align with your guidelines and always have TU costs.
2. A soldier skill menu which can activate actions that do not neccessarily have TU costs.
Would you approve of that?

File: src/Battlescape/ActionMenuState.cpp (hunk 3 above)
Comment (MeridianOXC) [reply src/Battlescape/ActionMenuState.cpp]: Adding something as an aimed shot just to "convert" it later to a snap shot by a script is just... stupid, sorry can't find a better word, no offense intended. I'd suggest to just keep adding defined actions one by one, until you reach 5 actions and then stop adding.

Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: There might be use cases for that scenario: SnapShot and AimedShot defined with different accuracy. SnapShot gets triggered on reaction fire, a skill triggers "50% chance to fire with aimed mode on reaction shots". The script can then throw the dice and decide wether to do a snap shot or an aimed shot.

File: src/Battlescape/TileEngine.cpp (hunk 4 above)
Comment (MeridianOXC) [reply src/Battlescape/TileEngine.cpp]: And how am I supposed to display this in Stats for Nerds, document in Ufopedia ruleset reference, check in the ruleset validator, describe on the forum, and support from now on until oxce death? It's basically unsupportable. Sigh.

File: src/Battlescape/BattlescapeGame.cpp (hunk 1 above)
Comment (MeridianOXC) [reply src/Battlescape/BattlescapeGame.cpp]: preliminary yes, final yes after I see it

File: src/Battlescape/ActionMenuState.cpp (hunk 3 above)
Comment (MeridianOXC) [reply src/Battlescape/ActionMenuState.cpp]: Yes, but it should be an exception, not a rule. The current "else if" logic would mean that everyone wanting to use a Use and Snapshot actions would need to make such a workaround with redefining actions. All I wanted to say is... get rid of the "else", add actions as long as you can (i.e. until you reach 5). Modders will understand the limit of 5 easier than the redefinition logic.

Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: Will do

File: src/Battlescape/ExplosionBState.cpp [hunk 5]
```diff
@@ -128,7 +132,7 @@ void ExplosionBState::init()
 				_power = 0;
 				miss = true;
 			}
-			else
+			else if (type != BT_SCRIPTED)
```
Comment (MeridianOXC) [inline src/Battlescape/ExplosionBState.cpp]: why is this here? no message to the user on panic/mc?

File: src/Battlescape/ExplosionBState.cpp [hunk 6]
```diff
@@ -110,16 +110,20 @@ void ExplosionBState::init()
...
 		}
 
 		//testing if we hit target
-		if (type == BT_PSIAMP && !_hit)
+		if ((type == BT_PSIAMP
+			 || (type == BT_SCRIPTED
+				 && (action == BA_MINDCONTROL || action == BA_PANIC))) && !_hit)
 		{
-			if (action != BA_USE)
+			if (action != BA_USE && type != BT_SCRIPTED)
```
Comment (MeridianOXC) [inline src/Battlescape/ExplosionBState.cpp]: why is this here? what does it mean?

File: src/Battlescape/ExplosionBState.cpp [hunk 7]
```diff
@@ -110,16 +110,20 @@ void ExplosionBState::init()
 			_damageType = itemRule->getDamageType();
 		}
 
-		if (type == BT_PSIAMP || _hit)
+		if ((type == BT_PSIAMP
+			 || (type == BT_SCRIPTED
+				  && (action == BA_MINDCONTROL || action == BA_PANIC))) || _hit)
```
Comment (MeridianOXC) [inline src/Battlescape/ExplosionBState.cpp]: why is BA_USE not allowed here?

File: src/Battlescape/ActionMenuState.cpp [hunk 8]
```diff
@@ -180,6 +180,56 @@ ActionMenuState::ActionMenuState(BattleAction *action, int x, int y) : _action(a
...
+		std::vector<SDLKey> hotKeys = {Options::keyBattleActionItem1, Options::keyBattleActionItem2, Options::keyBattleActionItem3, Options::keyBattleActionItem4, Options::keyBattleActionItem5};
+
+		if (!hotKeys.empty() && weapon->getCostUse().Time > 0)
+		{
+			addItem(BA_USE, "STR_USE_SCRIPTED", &id, hotKeys.back());
+			hotKeys.pop_back();
+		}
+		if (!hotKeys.empty() && weapon->getCostSnap().Time > 0)
+		{
+			addItem(BA_SNAPSHOT, "STR_SNAP_SCRIPTED", &id, hotKeys.back());
```
Comment (MeridianOXC) [inline src/Battlescape/ActionMenuState.cpp]: why are the default action names different? shouldn't the override be done per weapon?

File: src/Battlescape/ActionMenuState.cpp [hunk 9]
```diff
@@ -180,6 +180,56 @@ ActionMenuState::ActionMenuState(BattleAction *action, int x, int y) : _action(a
...
+		if (!hotKeys.empty() && weapon->getCostPrime().Time > 0)
+		{
+			addItem(BA_PRIME, "STR_PRIME_SCRIPTED", &id, hotKeys.back());
+			hotKeys.pop_back();
+		}
+		if (!hotKeys.empty() && weapon->getCostUnprime().Time > 0)
+		{
+			addItem(BA_UNPRIME, "STR_UNPRIME_SCRIPTED", &id, hotKeys.back());
+			hotKeys.pop_back();
+		}
```
Comment (MeridianOXC) [inline src/Battlescape/ActionMenuState.cpp]: is BA_LAUNCH missing by design or by accident?

File: src/Battlescape/ActionMenuState.cpp [hunk 10]
```diff
@@ -300,7 +358,7 @@ void ActionMenuState::btnActionMenuItemClick(Action *action)
 				_game->popState();
 			}
 		}
-		else if (_action->type == BA_UNPRIME)
+		else if (_action->type == BA_UNPRIME && weapon->getBattleType() != BT_SCRIPTED)
```
Comment (MeridianOXC) [inline src/Battlescape/ActionMenuState.cpp]: if prime and unprime are handled further below, shouldn't also BA_HIT be protected and handled below?

File: src/Battlescape/ActionMenuState.cpp [hunk 11]
```diff
@@ -465,6 +523,68 @@ void ActionMenuState::btnActionMenuItemClick(Action *action)
...
+				TileEngine *tileEngine = _game->getSavedGame()->getSavedBattle()->getTileEngine();
+				// is targeting action? (set _action->targeting = true, newHitLog = true)
+				// is melee action? newHitLog = true
+				// action response to the user? (set _action->result)
+				// is simple activation?
+				tileEngine->scriptedItemUse(_action);
+				
+				// action could have been changed by script now
+				switch (_action->type) {
+					case BA_PRIME: // default prime code
```
Comment (MeridianOXC) [inline src/Battlescape/ActionMenuState.cpp]: BA_UNPRIME missing or handled by the default clause?

File: src/Battlescape/ActionMenuState.cpp [hunk 12]
```diff
@@ -133,7 +133,7 @@ ActionMenuState::ActionMenuState(BattleAction *action, int x, int y) : _action(a
 		}
 	}
 
-	if (weapon->getCostMelee().Time > 0)
+	if (weapon->getCostMelee().Time > 0 && weapon->getBattleType() != BT_SCRIPTED)
```
Comment (MeridianOXC) [inline src/Battlescape/ActionMenuState.cpp]: should the same protection apply for the prime/unprime above?

File: src/Battlescape/ExplosionBState.cpp (hunk 5 above)
Comment (memmaker) [reply src/Battlescape/ExplosionBState.cpp]: The reason was, that I was using the scripted item as a method to have friendly/enemy remote targeting. So I wanted to have the targeting features of panic/mc but not the effect itself.

File: src/Battlescape/ExplosionBState.cpp (hunk 6 above)
Comment (memmaker) [reply src/Battlescape/ExplosionBState.cpp]: It just means, that the default code path for the PSI Amp is not being used by the scripted item.

File: src/Battlescape/ExplosionBState.cpp (hunk 7 above)
Comment (memmaker) [reply src/Battlescape/ExplosionBState.cpp]: Because BA_USE was used as a special untargeted action in this proposal.

File: src/Battlescape/ActionMenuState.cpp (hunk 11 above)
Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: Should be handled by default clause, iirc.

File: src/Battlescape/ActionMenuState.cpp (hunk 8 above)
Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: That was to ensure, that the action names wouldn't be confused with the default ones. Since the scripted actions can be completely different.

File: src/Battlescape/ActionMenuState.cpp (hunk 9 above)
Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: Bit of both, BA_LAUNCH was outside the scope.

File: src/Battlescape/ActionMenuState.cpp (hunk 10 above)
Comment (memmaker) [reply src/Battlescape/ActionMenuState.cpp]: BA_HIT should take the default code path and still work, that was the idea at least.
The switch statement below has a case for BA_HIT.
